### ISSUE TITLE
Aggregation cache files being regenerated for every request.

### ISSUE THREAD

**Issue by dblodgett-usgs** [Unidata/thredds#541]
I've really been chasing my tail on this one. I've stripped away all the variables and have identified what I believe to be the root problem of [Heiko's issue here.](http://www.unidata.ucar.edu/mailing_lists/archives/thredds/2015/msg00143.html)

To reproduce this issue, I 
1) created a fresh thredds 4.6 install. I tested 4.6.3 and 4.6.5. I let thredds 4.6.3 create the thredds content directory from scratch. 
2) I created a joinExisting aggregation of NetCDF4 files. ncml is:
```
<netcdf xmlns="http://www.unidata.ucar.edu/namespaces/netcdf/ncml-2.2">
	<aggregation dimName="time" type="joinExisting" timeUnitsChange="true">
		<scan location="files"/>
	</aggregation>
</netcdf>
```
3) I set up an as simple as possible catalog.
```
<?xml version="1.0" encoding="UTF-8"?>
<catalog xmlns="http://www.unidata.ucar.edu/namespaces/thredds/InvCatalog/v1.0" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="THREDDS Server Default Catalog : You must change this to fit your server!" xsi:schemaLocation="http://www.unidata.ucar.edu/namespaces/thredds/InvCatalog/v1.0            http://www.unidata.ucar.edu/schemas/thredds/InvCatalog.1.0.6.xsd">
	<service name="dap" base="" serviceType="compound">
		<service name="odap" serviceType="OpenDAP" base="/thredds/dodsC/"/>
	</service>
	<dataset name="Example of basic aggregation that rewrites cache files needlesly." ID="contrived.example" serviceName="dap" urlPath="example">
		<netcdf xmlns="http://www.unidata.ucar.edu/namespaces/netcdf/ncml-2.2" location="/Users/dblodgett/temp/thredds/data/example.ncml"/>
	</dataset>
</catalog>
```
4) I ran a script that makes requests for one time step for a small area in an iterative fashion--one request per time step. 

Every time I rerun this script, the cache files are rewritten. From what I can tell, they are being rewritten every time a request is made.

This has become a problem where the ascii cache files are very large. It is less noticeable for datasets with less time steps, but it is a very significant overhead both in terms of per-request latency and in terms of CPU to constantly rewrite (and presumably re-calculate those cache files). In the case of a [large aggregation I host here](http://cida.usgs.gov/thredds/dodsC/stageiv_combined.html), I see about 1 second per request (worse on servers with slower storage devices) which corresponds to writing two ~1MB cache files per request.

We do not experience this behavior prior to THREDDS 4.6. Typical latency per small request is less than 0.1 second, 20 requests per second is common. Unfortunately, the malformed xml in the .ddx response breaks an important application and other problems crop up in 4.5.3 and below, so I'm kind of stuck with 4.6+ but this performance issue is kind of a big deal. Is there possibly a workaround or a potential bug fix release to address it?

**Comment by cwardgar** [comment]
I cannot reproduce this. Here's what I have:
```
[cwardgar@lenny ~/dev/data/agg]
$ ls
total 112
drwxr-xr-x  16 cwardgar  ustaff   544B May 13 16:45 .
drwxr-xr-x  10 cwardgar  ustaff   340B Apr 30 12:34 ..
-rw-r--r--   1 cwardgar  ustaff   267B May 13 16:14 aggExisting.ncml
-rw-r--r--   1 cwardgar  ustaff   552B May 13 16:11 foo_2000-01-01.nc
-rw-r--r--   1 cwardgar  ustaff   552B May 13 16:11 foo_2000-01-02.nc
-rw-r--r--   1 cwardgar  ustaff   552B May 13 16:11 foo_2000-01-03.nc
-rw-r--r--   1 cwardgar  ustaff   552B May 13 16:11 foo_2000-01-04.nc
-rw-r--r--   1 cwardgar  ustaff   552B May 13 16:11 foo_2000-01-05.nc
-rw-r--r--   1 cwardgar  ustaff   552B May 13 16:11 foo_2000-01-06.nc
-rwxr--r--   1 cwardgar  ustaff   167B May 13 16:51 req_timesteps.sh
```
The NetCDF files look like:
```
[cwardgar@lenny ~/dev/data/agg]
$ ncdump foo_2000-01-01.nc
netcdf foo_2000-01-01 {
dimensions:
	time = 1 ;
	lat = 3 ;
	lon = 4 ;
variables:
	int time(time) ;
		time:units = "days since 2000-01-01" ;
	float lat(lat) ;
		lat:units = "degrees_north" ;
	float lon(lon) ;
		lon:units = "degrees_east" ;
	double temp(time, lat, lon) ;
		temp:units = "degC" ;
		temp:coordinates = "time lat lon" ;

// global attributes:
		:Conventions = "CF-1.6" ;
data:
 time = 0 ;
 lat = -60, 0, 60 ;
 lon = -90, -30, 30, 90 ;
 temp =
  1000, 1001, 1002, 1003,
  1004, 1005, 1006, 1007,
  1008, 1009, 1010, 1011 ;
}
```
The value of `time` is incremented for subsequent files.

Here's my simple NcML aggregation:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<netcdf xmlns="http://www.unidata.ucar.edu/namespaces/netcdf/ncml-2.2">
  <aggregation dimName="time" type="joinExisting" timeUnitsChange="true">
    <scan location="." regExp="\bfoo_.*\.nc$" />
  </aggregation>
</netcdf>
```
And a script I use to request each time value of the aggregation, one at a time:
```bash
#!/bin/bash
for i in `seq 0 5`;
do
    REQ="http://localhost:8080/thredds/dodsC/example.ascii?time[$i:1:$i]"
    echo "Requesting $REQ"
    curl $REQ
    sleep 1
done
```
My catalog is identical to yours, except with a different path to the NcML file.

After starting the server and running the script, I find a file named "example" in `<tds.content.root>/cache/agg/`:
```xml
<?xml version='1.0' encoding='UTF-8'?>
<aggregation xmlns='http://www.unidata.ucar.edu/namespaces/netcdf/ncml-2.2' version='3' type='joinExisting' dimName='time' >
  <netcdf id='/Users/cwardgar/dev/data/agg/foo_2000-01-01.nc' ncoords='1' >
    <cache varName='time' >0 </cache>
  </netcdf>
  <netcdf id='/Users/cwardgar/dev/data/agg/foo_2000-01-02.nc' ncoords='1' >
    <cache varName='time' >1 </cache>
  </netcdf>
  <netcdf id='/Users/cwardgar/dev/data/agg/foo_2000-01-03.nc' ncoords='1' >
    <cache varName='time' >2 </cache>
  </netcdf>
  <netcdf id='/Users/cwardgar/dev/data/agg/foo_2000-01-04.nc' ncoords='1' >
    <cache varName='time' >3 </cache>
  </netcdf>
  <netcdf id='/Users/cwardgar/dev/data/agg/foo_2000-01-05.nc' ncoords='1' >
    <cache varName='time' >4 </cache>
  </netcdf>
  <netcdf id='/Users/cwardgar/dev/data/agg/foo_2000-01-06.nc' ncoords='1' >
    <cache varName='time' >5 </cache>
  </netcdf>
</aggregation>
```
Running `stat`, I get:
```
[cwardgar@lenny ~/dev/tools/apache-tomcat-8.0.29/content/thredds/cache/agg]
$ stat -x example 
  File: "example"
  Size: 935          FileType: Regular File
  Mode: (0644/-rw-r--r--)         Uid: (23853/cwardgar)  Gid: ( 2000/  ustaff)
Device: 1,4   Inode: 6932213    Links: 1
Access: Fri May 13 17:07:31 2016
Modify: Fri May 13 16:21:31 2016
Change: Fri May 13 17:07:31 2016
```
After re-running the script, I do `stat` again:
```
$ stat -x example 
  File: "example"
  Size: 935          FileType: Regular File
  Mode: (0644/-rw-r--r--)         Uid: (23853/cwardgar)  Gid: ( 2000/  ustaff)
Device: 1,4   Inode: 6932213    Links: 1
Access: Fri May 13 17:20:33 2016
Modify: Fri May 13 16:21:31 2016
Change: Fri May 13 17:07:31 2016
```
Only the Access time has changed. It doesn't appear that the file was rewritten.

What's different between your setup and mine? I suspect it might only be an issue for large aggregations.

**Comment by dblodgett-usgs** [comment]
Well... I know I wasn't imagining things, but now I also can't reproduce this. My 4.6.5 deployment on cida-test is now just as fast as 4.5.5 cida production.  

My local on 4.6.5 with the simple example is also not reproducing the issue I was seeing before.  

My local on 4.6.3 with the simple example IS reproducing the issue.  

So, maybe my testing from above didn't actually get the switch from 4.6.3 to 4.6.5 completed. Really sorry for the run around. I guess back to #518 as the only blocker for us right now.

**Comment by dblodgett-usgs** [comment]
For the record, I've been working on some mongo nested aggregations for the MACAv2METDATA dataset that can be seen on our test server: http://cida-test.er.usgs.gov/thredds/catalog/macav2/catalog.html   

4.6.5 is performing way better for these nested joinexisting > union aggregations.

**Comment by cwardgar** [comment]
Don't sweat it; there are a lot of moving parts with these aggregations that make reproducing behavior difficult. I wouldn't be surprised if the underlying issue is still lurking somewhere, hidden from us for now.

Glad to hear about the performance increase; there was a caching fix in 4.6.5 that dramatically reduced the number of times that files in an aggregation dataset need to be reopened.
